### DISCUSSION TITLE
Disk space filling up causing SO to stop

### DISCUSSION THREAD

**Discussion by greatapoc** [Security-Onion-Solutions/securityonion#2615] [Unsupported Versions]
As referenced in https://github.com/Security-Onion-Solutions/securityonion/discussions/2581#discussioncomment-280493

Standalone Install - 2.3.21
1TB HDD
64GB RAM

My installation runs fine for a period of time until eventually the disk fills up and everything stops. This has been happening approximately once a week through every version I've been using and I've just put it down to my test machine being underspec but just wanted to raise the issue in case there's something else going on. I've always rectified it with a so-elastic-clear and start again.

Looking at the graphs it appears my disk space will constantly hover at 90% until eventually something breaks and it rapidly fills to 100% and grinds everything to a halt. Processes increase, CPU goes to 100%, ps aux shows a whole lot of stenoloss.sh defunct threads. I currently have it in this state so happy to provide any logs necessary.

![Capture1](https://user-images.githubusercontent.com/8947638/104513123-4415a980-5643-11eb-9369-6ca1d4901e7b.PNG)
![Capture2](https://user-images.githubusercontent.com/8947638/104513127-45df6d00-5643-11eb-875b-19de4514479c.PNG)

**Comment by weslambert** [comment]
Could this be the result of traffic bursting?  Does the throughput or EPS increase as you approach 100%?  Are there any clues in `/opt/so/log/sensor_clean.log` or `/opt/so/log/curator/curator.log`?

**Reply by greatapoc** [reply]
Looking at the graphs a bit narrower I've got monitor traffic going up to 500Mb/s flat from 23:00:00 GMT +11 (this was a backup job). The disk space started to spike up at 23:30:00. It regularly sees traffic over 500 up to 1Gb/s and doesn't seem to have an issue. Here's a wider graph of the previous 7 days.

![Capture3](https://user-images.githubusercontent.com/8947638/104515404-dff4e480-5646-11eb-9d1f-179850060cc8.PNG)

curator.log is 464MB. It is currently just being flooded with the following:
2021-01-13 21:52:09,842 ERROR     Failed to complete action: delete_indices.  <class 'curator.exceptions.FailedExecution'>: Failed to get indices. Error: TransportError(503, u'master_not_discovered_exception', None)
2021-01-13 21:53:05,744 INFO      Preparing Action ID: 1, "delete_indices"
2021-01-13 21:53:05,744 INFO      Creating client object and testing connection
2021-01-13 21:53:05,750 INFO      Instantiating client object
2021-01-13 21:53:05,751 INFO      Testing client connectivity
2021-01-13 21:53:05,758 INFO      Successfully created Elasticsearch client object with provided settings
2021-01-13 21:53:05,767 INFO      Trying Action ID: 1, "delete_indices": Delete indices when 651(GB) is exceeded.
2021-01-13 21:54:08,135 ERROR     Failed to complete action: close.  <class 'curator.exceptions.FailedExecution'>: Failed to get indices. Error: TransportError(503, u'master_not_discovered_exception', None)
2021-01-13 21:54:12,242 INFO      Preparing Action ID: 1, "close"
2021-01-13 21:54:12,242 INFO      Creating client object and testing connection
2021-01-13 21:54:12,251 INFO      Instantiating client object
2021-01-13 21:54:12,252 INFO      Testing client connectivity
2021-01-13 21:54:12,268 INFO      Successfully created Elasticsearch client object with provided settings
2021-01-13 21:54:12,282 INFO      Trying Action ID: 1, "close": Close Beats indices older than 30 days.
2021-01-13 21:55:05,805 ERROR     Failed to complete action: delete_indices.  <class 'curator.exceptions.FailedExecution'>: Failed to get indices. Error: TransportError(503, u'master_not_discovered_exception', None)
2021-01-13 21:56:06,376 INFO      Preparing Action ID: 1, "delete_indices"
2021-01-13 21:56:06,376 INFO      Creating client object and testing connection
2021-01-13 21:56:06,382 INFO      Instantiating client object
2021-01-13 21:56:06,383 INFO      Testing client connectivity
2021-01-13 21:56:06,393 INFO      Successfully created Elasticsearch client object with provided settings
2021-01-13 21:56:06,397 INFO      Trying Action ID: 1, "delete_indices": Delete indices when 651(GB) is exceeded.
2021-01-13 21:56:12,314 ERROR     Failed to complete action: close.  <class 'curator.exceptions.FailedExecution'>: Failed to get indices. Error: TransportError(503, u'master_not_discovered_exception', None)
2021-01-13 21:56:18,537 INFO      Preparing Action ID: 1, "close"
2021-01-13 21:56:18,538 INFO      Creating client object and testing connection
2021-01-13 21:56:18,543 INFO      Instantiating client object
2021-01-13 21:56:18,544 INFO      Testing client connectivity
2021-01-13 21:56:18,550 INFO      Successfully created Elasticsearch client object with provided settings
2021-01-13 21:56:18,557 INFO      Trying Action ID: 1, "close": Close Firewall indices older than 30 days.
2021-01-13 21:58:06,429 ERROR     Failed to complete action: delete_indices.  <class 'curator.exceptions.FailedExecution'>: Failed to get indices. Error: TransportError(503, u'master_not_discovered_exception', None)
2021-01-13 21:58:18,589 ERROR     Failed to complete action: close.  <class 'curator.exceptions.FailedExecution'>: Failed to get indices. Error: TransportError(503, u'master_not_discovered_exception', None)

/opt/so/log/sensor_clean.log just keeps filling up with "No old files available to clean up" for everything.

**Reply by weslambert** [reply]
Following back up on this -- whats is that state of the ES host? I notice 'master_not_discovered_exception', so I'm curious if this is a traditional ES cluster?  What does CURLing `https://localhost:9200/_cluster/health?pretty` provide you?

**Reply by greatapoc** [reply]
Sorry I did a so-elastic-clear to get it running again as I needed to use the system. This is a standalone install, no clustering.

**Reply by weslambert** [reply]
What is he output of the following?

`curl -s localhost:9200`
`curl -s localhost:9200/_cat/indices`
`curl -s localhost:9200/_cluster/health?pretty`

**Reply by greatapoc** [reply]
The system is currently operational. I'll redo these commands if/when it fills up again.

[root@securityonion onion]# curl -s localhost:9200
{
  "name" : "securityonion",
  "cluster_name" : "securityonion",
  "cluster_uuid" : "4cEHkQEQTUWEL7Wsu6wgzQ",
  "version" : {
    "number" : "7.9.3",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "c4138e51121ef06a6404866cddc601906fe5c868",
    "build_date" : "2020-10-16T10:36:16.141335Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
[root@securityonion onion]# curl -s localhost:9200/_cat/indices
green open .kibana-event-log-7.9.3-000002 eksmQ7weQlG2rq1KCjyuMg 1 0        8     0  43.1kb  43.1kb
green open .kibana-event-log-7.9.3-000001 19vbMMwNSaSuG0DnZ-M5zg 1 0       14     0  27.8kb  27.8kb
green open so-syslog-2021.01.19           eRbFk14LS7Gz_kMWg0zDFA 1 0   273186     0 349.2mb 349.2mb
green open elastalert_status              --Jy0j9MTcKO1-yZJzsgCw 1 0      581     0   2.7mb   2.7mb
green open .kibana-event-log-7.9.3-000003 01wZHvl2Tkiqzr6F9tsTow 1 0        4     0  21.6kb  21.6kb
green open so-strelka-2021.01.23          EDcD2_9TSJS2qUEF7iUi5g 1 0     2423     0     5mb     5mb
green open so-strelka-2021.01.22          TccCd-HLREazgwX5nT1ALg 1 0    85212     0   134mb   134mb
green open elastalert_status_silence      LvI7JyoIQFWnGP7WDIwI0g 1 0      586     0  93.4kb  93.4kb
green open so-strelka-2021.01.21          DvACxQxWReaYdHb4PTWG2w 1 0   107490     0 168.1mb 168.1mb
green open so-strelka-2021.01.20          wuA_slk3Qj-Q1Kp1Bq1sWg 1 0   116691     0 196.8mb 196.8mb
green open .apm-custom-link               JiWEn2ndQSyjHhNQFoWjqA 1 0        0     0    208b    208b
green open so-syslog-2021.01.20           xEBpxAQdQXalbGoefh7QzA 1 0   310374     0   395mb   395mb
green open so-zeek-2021.01.23             ofZq2lJOSSaWJyM9k9A-JQ 5 0  3468641     5   5.5gb   5.5gb
green open .kibana_task_manager_1         JrItMOAmTbWcY61eX9BjMQ 1 0        6 12183   3.7mb   3.7mb
green open so-ossec-2021.01.19            -CwTgRuwQhWw67sRpa0gyQ 1 0   623864     0   1.4gb   1.4gb
green open so-ids-2021.01.21              pudk5KSZQBCh6nDNeul1cg 1 0     2525     0  26.3mb  26.3mb
green open so-zeek-2021.01.20             AF8x3gztQ2CLraoOaD-aFg 5 0 57117398    19  80.4gb  80.4gb
green open so-ids-2021.01.20              dMxaD6MqQXmOozOnntYhVg 1 0     2127     0    11mb    11mb
green open so-zeek-2021.01.22             -MfVbgAFRdmCJGS4VBkdIA 5 0 75941537     8 116.3gb 116.3gb
green open so-ids-2021.01.23              x6zf1hbHSzijwBfshbRtJA 1 0       86     0   1.2mb   1.2mb
green open so-zeek-2021.01.21             Ds_grZQNTCSE9KCvuajCEw 5 0 69447008    17  99.1gb  99.1gb
green open so-ids-2021.01.22              GkOQSrBdSRigPrq8KFhLXA 1 0      752     0   4.8mb   4.8mb
green open elastalert_status_error        3V5bqR6aR-mHucp7Ke18hA 1 0        5     0  32.7kb  32.7kb
green open elastalert_status_status       pUZcydlRSF-TjcaBL_T-tw 1 0     3831     0 672.5kb 672.5kb
green open elastalert_status_past         ozSDp6ruTomgcASYmXRDAA 1 0        0     0    208b    208b
green open .apm-agent-configuration       h6ZZkxvuTriSCTN9jBZX9Q 1 0        0     0    208b    208b
green open so-syslog-2021.01.23           miVq84YQSwOS2SwsOMldnw 1 0      637     0   1.3mb   1.3mb
green open so-ossec-2021.01.20            PTSYVctbStaqaezg5hGgcw 1 0   178671     0 423.4mb 423.4mb
green open .kibana_5                      v4YtAjDUTiS7XmG6AkCRwA 1 0     2355    37  11.3mb  11.3mb
green open so-syslog-2021.01.21           rQCw7YqCR36BfTKpGmfgeQ 1 0   270885     0 338.2mb 338.2mb
green open so-syslog-2021.01.22           EHGkeLMyQSWbQErk2c1wtg 1 0   150537     0 187.9mb 187.9mb
green open .kibana_2                      ln9-W50qRC6gtzZfmMn0lQ 1 0     2437    47 930.7kb 930.7kb
green open so-ossec-2021.01.23            elzrAiOUTvaw6vQjLVVpjg 1 0    21712     0  56.3mb  56.3mb
green open .kibana_1                      ROZCH0MOSsmEmUdk0ZDRsQ 1 0     2221   211   1.1mb   1.1mb
green open .kibana_4                      _SIpBDYMS0uz79jSlgSurQ 1 0     2312    35 974.9kb 974.9kb
green open so-ossec-2021.01.22            oOzEor6JRW2MN2NOkrdBfQ 1 0   225563     0 547.6mb 547.6mb
green open so-ossec-2021.01.21            Hz6p7NYdS8y4pLw72_bSBg 1 0   158394     0 385.7mb 385.7mb
green open .kibana_3                      gCqrCo-HRGODt_UF3LCbaA 1 0     2302    35 923.6kb 923.6kb
green open .tasks                         SX7dZPmTTNuaxJXD8N2YiQ 1 0        4     0  20.7kb  20.7kb
green open so-ids-2021.01.19              dsB6C2wAQEeDSL7MGxONMA 1 0     1748     0   9.4mb   9.4mb
green open .async-search                  uaYQyGL5QVaiDuzg6brC8g 1 0        0     0   4.8kb   4.8kb
[root@securityonion onion]# curl -s localhost:9200/_cluster/health?pretty
{
  "cluster_name" : "securityonion",
  "status" : "green",
  "timed_out" : false,
  "number_of_nodes" : 1,
  "number_of_data_nodes" : 1,
  "active_primary_shards" : 60,
  "active_shards" : 60,
  "relocating_shards" : 0,
  "initializing_shards" : 0,
  "unassigned_shards" : 0,
  "delayed_unassigned_shards" : 0,
  "number_of_pending_tasks" : 0,
  "number_of_in_flight_fetch" : 0,
  "task_max_waiting_in_queue_millis" : 0,
  "active_shards_percent_as_number" : 100.0
}

**Reply by weslambert** [reply]
How many iterations of upgrades/soups has this box been through?  Looks like there are quite a few Kibana indices, so curious if it could have been upgraded through several Elastic versions.

**Reply by greatapoc** [reply]
This has been souped up every step from 2.3 GA. I think I'll just do a full rebuild when the next version comes out.

**Reply by weslambert** [reply]
Please try modifying `log_size_limit` to be half of the disk space (in GB) available for `/nsm` in `/opt/so/saltstack/local/pillar/minions/$nodename_standalone.sls` and see if you continue to run into the same issue.  Also, please let me know the size of `/nsm/strelka/log/strelka.log`.

Thanks,
Wes

**Reply by greatapoc** [reply]
It filled up overnight again. No burst of traffic this time.
![Capture23](https://user-images.githubusercontent.com/8947638/106054142-ba380700-613f-11eb-8673-b90f13786e56.PNG)
![Capture24](https://user-images.githubusercontent.com/8947638/106054154-bc01ca80-613f-11eb-9a8f-96cd8819ac97.PNG)

[root@securityonion onion]# curl -s localhost:9200
{
  "name" : "securityonion",
  "cluster_name" : "securityonion",
  "cluster_uuid" : "4cEHkQEQTUWEL7Wsu6wgzQ",
  "version" : {
    "number" : "7.9.3",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "c4138e51121ef06a6404866cddc601906fe5c868",
    "build_date" : "2020-10-16T10:36:16.141335Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"

[root@securityonion onion]# curl -s localhost:9200/_cat/indices
{"error":{"root_cause":[{"type":"master_not_discovered_exception","reason":null}],"type":"master_not_discovered_exception","reason":null},"status":503}

[root@securityonion onion]# curl -s localhost:9200/_cluster/health?pretty
{
  "error" : {
    "root_cause" : [
      {
        "type" : "master_not_discovered_exception",
        "reason" : null
      }
    ],
    "type" : "master_not_discovered_exception",
    "reason" : null
  },
  "status" : 503
}

[root@securityonion log]# du -hsx *
16G     strelka.log
36K     yara-update.log

**Reply by weslambert** [reply]
After looking at this with the rest of the team, it seems there could be a few things going on. You'll want to reduce log_size_limit (/opt/so/saltstack/loca/pillar/minions/$node.sls) to 50% of available disk space (in GB) for /nsm. Additionally, you may need to check the size of the Strelka log in /nsm/strelka/log , removing it and restarting Strelka if necessary (can get quite large -- we are working on rotation). Last, you may have quite a few so-sensor-clean processes running. So, you may need to modify the so-sensor-clean script as detailed here: 9448177 in order to prevent it self from running multiple instances, then clean up the existing ones, if necessary (find them with ps aux | grep sensor-clean | grep -v grep, then kill the processes, or reboot).

**Reply by greatapoc** [reply]
So I reduced the log_size_limit and it has just now hit that limit and stopped taking any additional logs. I increased it by 50GB and restarted elasticsearch and it has started reading new logs again. It seems like it is a state of balance taking space from PCAPs and so on but what function is there to rotate Elasticsearch Documents once the log limit is reached? Is it a time based thing? I would've expected it to operate as FIFO once that limit is reached.

**Reply by weslambert** [reply]
Yes, so depending on the amount of space available to the partition on which /nsm resides, that value will limit should be set to around 50% of the available space when running a standalone.  This limits Elasticsearch in writing data, so it does not corrupt something.  There was a bug in this most recent release and prior that did not calculate this percentage correctly (calculated for the entire disk, and not just the partition, and there were also a few other things that needed cleaned up, like the so-sensor-clean script running multiple instances, and the Strelka log being able to grow very large in size).  It sounds like other files are causing it to realize that value rather immediately.  The sensor cleanup job does run every minute and deletes certain files  like Zeek logs, Strelka files, etc (and Steno cleans itself up), but there may be some conflict there.  Do you notice any unusual amount of so-sensor-clean processes, or any large files (like the Strelka log) within /nsm, other than the ES data?
